File: challenges-in-python/two_pointers/palindrome.py
# Big O: O(n) time, O(n) space
def is_palindrome(s: str) -> bool:
    cleaned = ''.join([c for c in s.lower() if c.isalnum()])
    return cleaned == cleaned[::-1]

# Big O: O(n) time, O(1) space
def is_palindrome(s: str) -> bool:
    # check for empty string
    if not s: return True
    left, right = 0, len(s) - 1
    while left < right:
        # skip non-alphanumeric characters
        while left < right and not s[left].isalnum(): left += 1
        while left < right and not s[right].isalnum(): right -=1

        # Check if charecters are equal
        if s[left].lower() != s[right].lower():
            return False
        
        left += 1
        right -= 1

    return True

assert(is_palindrome("A man, a plan, a canal, Panama"))
assert(is_palindrome(""))
assert(is_palindrome("a"))
assert not (is_palindrome('hello'))
assert(is_palindrome('racecar'))

File: challenges-in-rust/src/two_pointers/palindrome.rs
// Checking for Palindrome
// A palindrome is a string that reads the same forward and backward. 

// Big O Analysis:
// Time complexity: O(n) | Space complexity: O(n)
pub fn is_palindrome(s: &str) -> bool {
    let s: Vec<char> = s.chars().filter(|c: &char| c.is_alphanumeric()).collect();
    let mut left: usize = 0;
    let mut right: usize = s.len().checked_sub(1).unwrap_or(0);
    
    while left < right {
        if s[left].to_lowercase().ne(s[right].to_lowercase()) {
            return false;
        }
        left += 1;
        right -= 1;
    }
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_palindrome_simple() {
        assert!(is_palindrome("racecar"));
        assert!(!is_palindrome("rust"));
    }

    #[test]
    fn test_is_palindrome_complex() {
        assert!(is_palindrome("A man, a plan, a canal: Panama"));
        assert!(!is_palindrome("This is not a palindrome"));
    }

    #[test]
    fn test_is_palindrome_case_insensitive() {
        assert!(is_palindrome("RaceCar"));
        assert!(is_palindrome("No lemon, no melon"));
    }

    #[test]
    fn test_is_palindrome_empty_and_single_char() {
        assert!(is_palindrome(""));
        assert!(is_palindrome("a"));
    }
}